### PR TITLE
{math}[foss-2016a] Octave 4.0.0 (REVIEW)

### PR DESCRIPTION
**Pull request by hajgato**
needs easyblock https://github.com/hpcugent/easybuild-easyblocks/pull/863

PR ~~#2743 #2744 #2745 #2746 #2752 #2753 #2754 #2756 #2757 #2758 #2759 #2760 #2761 #2763~~
edit: requires ~~https://github.com/hpcugent/easybuild-easyblocks/pull/863~~

### CODE REVIEW

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 4 (4 easyconfigs in this PR)
node2604.swalot.os - Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz, Python 2.7.5
See https://gist.github.com/f15139da9ad3d250b374072c4a5088fe for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 4 (4 easyconfigs in this PR)
node2159.delcatty.os - Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/826e12d80d630bbcfef4c053dd256862 for a full test report.

**Comment by boegel** [comment]
@hajgato installation is failing for me for two different reasons:

```
== 2016-04-09 18:52:06,269 build_log.py:152 ERROR Sanity check failed: sanity check command octave --eval "1+2" exited with code 127 (output: octave: X11 DISPLAY environment variable not set
octave: disabling GUI features
/user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/CO7/haswell-ib/software/Octave/4.0.0-foss-2016a/bin/octave-cli-4.0.0:
error while loading shared libraries: libcholmod.so.3: cannot open shared object file: No such file or directory
```

and

```
src/moc-dialog.cc:17:2: error: #error "This file was generated using the moc from 4.6.2. It"
 #error "This file was generated using the moc from 4.6.2. It"
  ^
src/moc-dialog.cc:18:2: error: #error "cannot be used with the include files from this version of Qt."
 #error "cannot be used with the include files from this version of Qt."
  ^
src/moc-dialog.cc:19:2: error: #error "(The moc has changed too much.)"
 #error "(The moc has changed too much.)"
  ^
```

I'm not sure why I'm not seeing the same problem in both instances...

**Comment by hajgato** [comment]
@boegel:  The first: needs `SuiteSparse`  new easyblock. https://github.com/hpcugent/easybuild-easyblocks/pull/863 
And I have to investigate the second. (For some reason it finds Qt 4.6.2 (which is probably system installed), instead of Qt 4.8.7.

**Comment by hajgato** [comment]
@boegel: for the second error I have to see when config picked up the system `Qt`

**Comment by hajgato** [comment]
@boegel: is it possible to get the full log for `node2159.delcatty.os - Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6`

**Comment by boegel** [comment]
@hajgato I'll create a gist for it and get back to you

**Comment by boegel** [comment]
@hajgato https://gist.github.com/boegel/dd1ef8fb49e6ff443708534e496d9aa4

**Comment by hajgato** [comment]
@boegel: strange, it found `Qt-4.8.7`, but for some reason it uses `moc` from `Qt-4.6.2`. Can you check whether the file `$EBROOTQT/bin/moc` exisit?

another possibility that the system `qmake` is used instead of the EB installed one, but I am not able to check this from the log.

**Comment by hajgato** [comment]
@boegel: and can you check `$EBROOTQT/bin/qmake` as well?

File: easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+sources = [SOURCE_TAR_GZ]
+
+metis = 'METIS'
+metis_ver = '5.1.0'
+versionsuffix = '-%s-%s' % (metis, metis_ver)
+dependencies = [(metis, metis_ver)]
+
+preconfigopts = 'LAPACK="$LIBLAPACK" ' 
+prebuildopts = 'LAPACK="$LIBLAPACK" ' 
+preinstallopts = 'LAPACK="$LIBLAPACK" '
```
Comment (boegel) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @hajgato the updated easyblock in https://github.com/hpcugent/easybuild-easyblocks/pull/863 should handle this, no?

cc @besserox, @jrbosch

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: This is not handled by the SuiteSparse easyblock. Probably it should.

From where is coming the `LIBLAPACK` environment variable?
What is the content of the variable? Does it work with any toolchain?

According to the README.txt of SuiteSparse, we can (should) also set `BLAS=...`

Comment (boegel) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @besserox it's defined by the EasyBuild toolchain mechanism, and will always be defined if a BLAS/LAPACK library is included in the toolchain; see http://easybuild.readthedocs.io/en/latest/Extended_dry_run_examples.html#extended-dry-run-examples for examples

there's also a `$LIBBLAS` equivalent

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: For reference, an extract of README.txt of SuiteSparse 4.5.1
```
(1) Use the right BLAS and LAPACK libraries

    Determine where your BLAS and LAPACK libraries are.  If the default
    'make' does not find them, use
    'make BLAS=-lmyblaslibraryhere LAPACK=-lmylapackgoeshere'
```

Comment (boegel) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @besserox that's exactly what `$LIBBLAS` and `$LIBLAPACK` are intended for

Comment (hajgato) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @boegel I think yes. It would be nice. 
cc @besserox, @jrbosch

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: I will try to add this to #863, unless you prefer a separate PR

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: It appears to be already there:
https://github.com/besserox/easybuild-easyblocks/blob/b086deb874af6322a33936ba321d8010a6a1949d/easybuild/easyblocks/s/suitesparse.py#L71
```
             'BLAS': os.getenv('LIBBLAS_MT'),
              'LAPACK': os.getenv('LIBLAPACK_MT'),
```
@hajgato Why did you add these lines initially? Did you notice SuiteSparse was not picking the right LAPACK?

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @boegel @hajgato @jrbosch
After a quick check, it appears the SuiteSparse libraries are linked to the right BLAS/LAPACK. I ran `ldd libumfpack.so` (which is built by SuiteSparse) and it is correctly linked to OpenBLAS or IntelMLK respectively if I use `foss` or `ictce` toolchain.

So, there is no need to change #863 and these 3 lines can be safely removed from the easyconfig file:
```
preconfigopts = 'LAPACK="$LIBLAPACK" ' 
prebuildopts = 'LAPACK="$LIBLAPACK" ' 
preinstallopts = 'LAPACK="$LIBLAPACK" ' 
```

Comment (boegel) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @jrbosch mentioned in #2839 that he ran into build problems without these using `SuiteSparse-4.5.2-foss-2016a-METIS-5.1.0.eb`

```
@boegel SuiteSparse easyconfig was taked from PR #2766 and I just add MY_METIS_LIB.
If I don't include 'LAPACK="$LIBLAPACK" one error occur in building step because lapack.
```

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: After a second check, it appears the library is linked to both `libopenblas.so` and `libblas.so`.
```
$ ldd $EBROOTSUITESPARSE/lib/libumfpack.so
...
        liblapack.so.3 => /usr/lib/liblapack.so.3 (0x00007f2595e63000)
        libopenblas.so.0 => /home/users/xbesseron/.local/easybuild/software/numlib/OpenBLAS/0.2.15-GCC-4.9.3-2.25-LAPACK-3.6.0/lib/libopenblas.so.0 (0x00007f25951c5000)
        libblas.so.3 => /usr/lib/libblas.so.3 (0x00007f2594416000)
...
```

This happens only for `foss`. It appears to be OK for `ictce`.

Comment (hajgato) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @besserox probably depends on whether system lapack and bals are installed. We have to prevent to link against system lapack/blas.

Comment (boegel) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @besserox let's open an issue for this and handle it via a new PR for the SuiteSparse easyblock...

https://github.com/hpcugent/easybuild-easyblocks/pull/863 is ready to go, I'd like to include it in EB v2.8.0

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: I think I got it right now.

PR https://github.com/hpcugent/easybuild-easyblocks/pull/863 updates the SuiteSparse easyblock to set BLAS and LAPACK properly on the build command line.

So you should be able to remove this 3 lines.

Comment (besserox) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @hajgato You can remove these 3 lines. PR hpcugent/easybuild-easyblocks#863 will take care of defining these variables properly.

Comment (hajgato) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @besserox I will do that.

Comment (jrbosch) [inline easybuild/easyconfigs/s/SuiteSparse/SuiteSparse-4.5.1-foss-2016a-METIS-5.1.0.eb]: @besserox @hajgato @boegel I see that the problem will be solved. but just in case I leave here part of my log when I tried to install SuiteSparse without setting LAPACK variable:

gcc -fPIC -funroll-loops -O3 -march=native 
-I/home/CLUSTER/uclv.jrbosch/.local/easybuild/software/OpenBLAS/0.2.15-GCC-4.9.3-2.25-LAPACK-3.6.0/include 
-I/home/CLUSTER/uclv.jrbos$
gcc -fPIC -funroll-loops -O3 -march=native 
-I/home/CLUSTER/uclv.jrbosch/.local/easybuild/software/OpenBLAS/0.2.15-GCC-4.9.3-2.25-LAPACK-3.6.0/include 
-I/home/CLUSTER/uclv.jrbos$
gcc -fPIC -funroll-loops -O3 -march=native 
-I/home/CLUSTER/uclv.jrbosch/.local/easybuild/software/OpenBLAS/0.2.15-GCC-4.9.3-2.25-LAPACK-3.6.0/include 
-I/home/CLUSTER/uclv.jrbos$
./cholmod_demo < Matrix/bcsstk01.tri
./cholmod_demo: symbol lookup error: /usr/lib/liblapack.so.3: undefined symbol: gotoblas
Makefile:31: recipe for target 'all' failed
make[2]: *** [all] Error 127
make[2]: Leaving directory 
'/home/CLUSTER/uclv.jrbosch/.local/easybuild/build/SuiteSparse/4.5.2/foss-2016a-METIS-5.1.0/SuiteSparse/CHOLMOD/Demo'
Makefile:14: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory 
'/home/CLUSTER/uclv.jrbosch/.local/easybuild/build/SuiteSparse/4.5.2/foss-2016a-METIS-5.1.0/SuiteSparse/CHOLMOD'
Makefile:14: recipe for target 'go' failed
make: *** [go] Error 2
 (at 
easybuild/software/EasyBuild/2.7.0/lib/python2.7/site-packages/easybuild_framework-2.7.0-py2.7.egg/easybuild/tools/run.
py:397 in parse_cmd_output)
== 2016-05-11 23:56:54,700 easyblock.py:2296 WARNING build failed (first 300 chars): cmd 
"MY_METIS_LIB="$EBROOTMETIS/lib/libmetis.so"  make -j 1 " exited with exitcode 2 and ou$
Using pre-installed METIS 5.1.0 library at  
[/home/CLUSTER/uclv.jrbosch/.local/easybuild/software/METIS/5.1.0-foss-2016a/lib/libmetis.so]
( cd SuiteSparse_config && make )
make[1]: Entering directory '/h
== 2016-05-11 23:56:54,700 easyblock.py:267 INFO Closing log for application name SuiteSparse version 4.5.2

**Comment by hajgato** [comment]
@boegel What to do with this? I have no idea why the build does not work for you, other than it finds ` moc` from `Qt 4.6.2` (probably a repository installed version instead `moc` from the EB provided `Qt 4.8.7`

**Comment by boegel** [comment]
@hajgato both `moc` and `qmake` are there:

```
$ ls $EASYBUILD_PREFIX/software/Qt/4.8.7-foss-2016a/bin
assistant  lconvert  lrelease  moc	  qcollectiongenerator	qhelpconverter	qmake	       qmlviewer  qtconfig  qttracereplay  uic	 xmlpatterns
designer   linguist  lupdate   pixeltool  qdoc3			qhelpgenerator	qmlplugindump  qt3to4	  qtdemo    rcc		   uic3  xmlpatternsvalidator
```

Do we need to tell Octave the full path to `qmake` and/or `moc` to use? I've seen that before...

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 4 (4 easyconfigs in this PR)
node2159.delcatty.os - Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/49d579af4ed70249cd7467c57a7329dc for a full test report.

**Comment by hajgato** [comment]
@boegel The problem was that configure searches for `moc` in the following order: `moc-qt4`,`moc`. The EB isnatlled `Qt` have no `moc-qt4`, only `moc`. So If you have a  repository installed `Qt4` with `moc-qt4`, the system will be used. We have to remember this problem, because bote EB `Qt4` and `Qt5` uses `moc` and none of them `moc-qtX`

File: easybuild/easyconfigs/o/Octave/Octave-4.0.0-foss-2016a.eb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    ('libGLU', '9.0.0'),
+    ('Qhull', '2015.2'),
+    ('Qt', '4.8.7'),
+    ('HDF5', '1.8.16', '-serial'),
+    ('qrupdate', '1.1.2'),
+    ('SuiteSparse', '4.5.1', '-METIS-5.1.0'),
+    ('GraphicsMagick', '1.3.23'),
+]
+
+configopts = 'MOC=$EBROOTQT/bin/moc --with-blas="$LIBBLAS" --with-lapack="$LIBLAPACK" '
```
Comment (boegel) [inline easybuild/easyconfigs/o/Octave/Octave-4.0.0-foss-2016a.eb]: @hajgato should this be this instead?

```python
preconfigopts = "export MOC=$EBROOTQT/bin/moc && "
```

Comment (hajgato) [inline easybuild/easyconfigs/o/Octave/Octave-4.0.0-foss-2016a.eb]: For me it is fine. but I think configure understands both (for example `./configure CC=pgcc`)
But if you think that `preconfigopts = "export MOC=$EBROOTQT/bin/moc && "` is better, I will use that.

Comment (hajgato) [inline easybuild/easyconfigs/o/Octave/Octave-4.0.0-foss-2016a.eb]: @boegel ^

Comment (boegel) [inline easybuild/easyconfigs/o/Octave/Octave-4.0.0-foss-2016a.eb]: @hajgato well, if both work, this is fine by me, I'll retest this PR

**Comment by boegel** [comment]
@hajgato we can install a symlink in the Qt easyconfigs to overcome this issue? Since we have an easyblock for Qt, this is quite easy to do?

File: easybuild/easyconfigs/o/Octave/Octave-4.0.0-foss-2016a.eb (hunk 2 above)
Comment (hajgato) [inline easybuild/easyconfigs/o/Octave/Octave-4.0.0-foss-2016a.eb]: @boegel from  https://www.gnu.org/prep/standards/html_node/Configuration.html
Specifying variables as arguments to configure, like this:
```
./configure CC=gcc
```
is preferable to setting them in environment variables:
```
CC=gcc ./configure
```
as it helps to recreate the same configuration later with config.status. However, both methods should be supported.

**Comment by hajgato** [comment]
@boegel I think yes. We should check what other `Qt` commands have the `-qtX` suffix. On the other hand, it might be distro specific.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 4 (4 easyconfigs in this PR)
node2159.delcatty.os - Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/1c7160fe5fbea6dab6e54238b8a81d60 for a full test report.

**Comment by boegel** [comment]
@hajgato hmm, now the installation works fine, but the (very simple) sanity check fails?

**Comment by hajgato** [comment]
@boegel Look, I am giving up on this project. The problem is FLTK, and most probably due to missing X-org headers as builddep. We have to once an all fix the x-org misery we have with 20 different headers and libs. I will not touch any x-org related project until we solve #2579 and the related easyblock https://github.com/hpcugent/easybuild-easyblocks/pull/843

**Comment by hajgato** [comment]
@boegel The bottom line is that for me `FLTK` works, and for you not. I think it is clearly an `FLTK` related problem, and I am not able to debug it.

**Comment by boegel** [comment]
@hajgato is bumping this to `foss/2016b` and using the `X11` an option to make this work?

**Comment by hajgato** [comment]
@boegel Maybe. I will work on it.

**Comment by hajgato** [comment]
@boegel: It seems that I found the problem. `Qt/moc` issues are hopefully solved, and the `FLTK` issue is solved by PR #3704 (`FLTK` have to be rebuild, a patch is applied)

**Comment by fgeorgatos** [comment]
good job here...

**Comment by boegel** [comment]
lgtm

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 59 out of 60 (3 easyconfigs in this PR)
node2077.delcatty.os - Linux SL 6.8, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/bff181bb7f6f448a4c61afce7bd3ebfa for a full test report.

**Comment by boegel** [comment]
@hajgato any idea why I'm getting this for the Octave sanity check, but not for #3717? Is this because of the different Octave version?

```
octave: X11 DISPLAY environment variable not set
```

**Comment by boegel** [comment]
@hajgato actually, the problem is a missing dep, see https://github.com/hpcugent/easybuild-easyconfigs/pull/3717#issuecomment-257834225

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
node2301.phanpy.os - Linux centos linux 7.2.1511, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz, Python 2.7.5
See https://gist.github.com/eac496be989e22b0f6b9cd49c9daa61f for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (3 easyconfigs in this PR)
node1001.muk.os - Linux SL 6.8, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/b3c2e7bf22e21e1733362412868a1443 for a full test report.

**Comment by boegel** [comment]
Going in, thanks @hajgato!
